server/src/main/java/org/opensearch/action/support/replication/ReplicationOperation.java

```diff
@@ -226,10 +233,158 @@ private void performOnReplicas(
 
         final ShardRouting primaryRouting = primary.routingEntry();
 
-        for (final ShardRouting shard : replicationGroup.getReplicationTargets()) {
-            if (shard.isSameAllocation(primaryRouting) == false) {
-                performOnReplica(shard, replicaRequest, globalCheckpoint, maxSeqNoOfUpdatesOrDeletes, pendingReplicationActions);
+        for (final ReplicationModeAwareShardRouting shardRouting : replicationGroup.getReplicationTargets()) {
+            replicationProxy.performOnReplica(
+                shardRouting,
+                primaryRouting,
+                replicaRequest,
+                globalCheckpoint,
+                maxSeqNoOfUpdatesOrDeletes,
+                pendingReplicationActions
+            );
+        }
+    }
+
+    /**
+     * Factory that return the {@link ReplicationProxy} instance basis the {@link ReplicationModeAwareOverrideProxy}.
+     *
+     * @opensearch.internal
+     */
+    private class ReplicationProxyFactory {
```

Comment:
Can probably just make this a static method since it has no state and the created instance is thrown away after use?

Reply:
I see. Just an instance method would be preferable, in my opinion.

Reply:
Yes, hence i have kept it as instance method only.